crates/broker/src/storage.rs

```diff
@@ -2,296 +2,311 @@
+        }
 
-    fn supported_bonsai_host(authority: &str) -> bool {
-        matches!(authority, "image" | "input")
+        Ok(tokio::fs::read(&self.path).await?)
     }
+}
+
+pub struct HttpHandler {
+    url: url::Url,
+    client: ClientWithMiddleware,
+    max_size: usize,
+}
 
-    pub fn new(
-        uri_str: &str,
-        max_size: Option<usize>,
-        retries: Option<u8>,
+impl HttpHandler {
+    async fn new(
+        url: url::Url,
+        max_size: usize,
         cache_dir: Option<PathBuf>,
+        max_retries: Option<u8>,
     ) -> Result<Self, StorageErr> {
-        let uri = url::Url::parse(uri_str)?;
-
-        let scheme = uri.scheme().to_string();
-
-        if !Self::supported_scheme(&scheme) {
-            return Err(StorageErr::UnsupportedScheme(scheme));
+        if !matches!(url.scheme(), "http" | "https") {
+            return Err(StorageErr::InvalidURL("invalid HTTP scheme"));
         }
-
-        // file scheme is only supported in dev mode
-        if scheme == "file" && !risc0_zkvm::is_dev_mode() {
-            return Err(StorageErr::UnsupportedScheme(scheme));
+        if !url.has_host() {
+            return Err(StorageErr::InvalidURL("missing host"));
         }
 
-        if scheme == "bonsai" {
-            let authority = http::uri::Authority::from_str(uri.authority())
-                .map_err(|_| StorageErr::AuthorityMissing)?;
-            if !Self::supported_bonsai_host(authority.host()) {
-                return Err(StorageErr::InvalidBonsaiHost(authority.host().to_string()));
-            }
-
-            let path = uri.path();
-            if path.is_empty() || path == "/" {
-                return Err(StorageErr::NullBonsaiPath);
-            }
-        }
+        let mut builder = ClientBuilder::new(reqwest::Client::new());
 
-        let client = if let Some(cache_dir) = cache_dir {
+        if let Some(cache_dir) = cache_dir {
+            tokio::fs::create_dir_all(&cache_dir).await?;
             let manager = CACacheManager { path: cache_dir };
-            let cache = Cache(HttpCache {
+            let cache_middleware = Cache(HttpCache {
                 mode: CacheMode::ForceCache,
                 manager,
                 options: HttpCacheOptions::default(),
             });
-            ClientBuilder::new(reqwest::Client::new()).with(cache).build()
-        } else {
-            ClientBuilder::new(reqwest::Client::new()).build()
-        };
 
-        Ok(Self {
-            uri,
-            uri_scheme: scheme,
-            max_size,
-            retries: retries.unwrap_or(DEFAULT_RETRY_NUMB),
-            client,
-        })
-    }
+            builder = builder.with(cache_middleware)
+        }
+        if let Some(max_retries) = max_retries {
+            let retry_policy =
+                ExponentialBackoff::builder().build_with_max_retries(max_retries as u32);
+            let retry_middleware = RetryTransientMiddleware::new_with_policy(retry_policy);
```

Comment:
Previously we had implemented manual retry logic that only retried the initial get state. While the `RetryTransientMiddleware` is much nicer, it only exits on transient errors, whereas before we didn't have that distinction and exited on everything. Conceptually, I think retiring only on transient errors should be the right choice.